### ISSUE TITLE
:TSInstall while committing may break current git repo

### ISSUE THREAD

**Issue by staticssleever668** [nvim-treesitter/nvim-treesitter#5353]
### Describe the bug

Running `:TSInstall gitcommit` while committing in git broke my repo.
I use `require("nvim-treesitter.install").prefer_git = true`.

### To Reproduce

After encountering it on remote server, I immediately retried and reproduced it once on my home PC, but on subsequent attempts I can't reproduce it again.
Rough steps:
```
mkdir ~/Downloads/test
cd ~/Downloads/test
git init
touch asdf
git add asdf
git commit --no-gpg-sign
nvim asdf # ijklsdfjklsd:x
git add asdf
git commit # ikljsadflkjsfadjlk<ESC>:TSInstall gitcommit:x
git status
# > see error "fatal: unable to read 11544a188ec3192eb3aab635520020d5f175bd0b"
```

### Output of `:checkhealth nvim-treesitter`

Initial encounter:
```
==============================================================================
nvim-treesitter: require("nvim-treesitter.health").check()

Installation ~
- WARNING `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)
- OK `node` found v12.22.9 (only needed for :TSInstallFromGrammar)
- OK `git` executable found.
- OK `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
  Version: cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
- OK Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

OS Info:
{
  machine = "x86_64",
  release = "6.2.0-31-generic",
  sysname = "Linux",
  version = "#31~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Aug 16 13:45:26 UTC 2"
} ~

Parser/Features         H L F I J
  - bash                ✓ ✓ ✓ . ✓
  - c                   ✓ ✓ ✓ ✓ ✓
  - cmake               ✓ . ✓ ✓ .
  - cpp                 ✓ ✓ ✓ ✓ ✓
  - lua                 ✓ ✓ ✓ ✓ ✓
  - markdown            ✓ . ✓ ✓ ✓
  - markdown_inline     ✓ . . . ✓
  - python              ✓ ✓ ✓ ✓ ✓
  - query               ✓ ✓ ✓ ✓ ✓
  - sql                 ✓ . . ✓ ✓
  - vim                 ✓ ✓ ✓ . ✓
  - vimdoc              ✓ . . . ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang} ~
```

Reproduced:
```
==============================================================================
nvim-treesitter: require("nvim-treesitter.health").check()

Installation ~
- WARNING `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)
- OK `node` found v20.6.0 (only needed for :TSInstallFromGrammar)
- OK `git` executable found.
- OK `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
  Version: cc (GCC) 13.2.1 20230801
- OK Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

OS Info:
{
  machine = "x86_64",
  release = "6.4.12-arch1-1",
  sysname = "Linux",
  version = "#1 SMP PREEMPT_DYNAMIC Thu, 24 Aug 2023 00:38:14 +0000"
} ~

Parser/Features         H L F I J
  - c                   ✓ ✓ ✓ ✓ ✓
  - cpp                 ✓ ✓ ✓ ✓ ✓
  - glsl                ✓ ✓ ✓ ✓ ✓
  - java                ✓ ✓ ✓ ✓ ✓
  - javascript          ✓ ✓ ✓ ✓ ✓
  - json5               ✓ . . . ✓
  - lua                 ✓ ✓ ✓ ✓ ✓
  - python              ✓ ✓ ✓ ✓ ✓
  - rust                ✓ ✓ ✓ ✓ ✓
  - toml                ✓ ✓ ✓ ✓ ✓
  - xml                 ✓ . ✓ ✓ ✓
  - yaml                ✓ ✓ ✓ ✓ ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang} ~
```

### Output of `nvim --version`

Initial encounter:
```text
NVIM v0.10.0-dev-1031+gc431d820e
Build type: RelWithDebInfo
LuaJIT 2.1.1693350652
Run "nvim -V1 -v" for more info
```

Reproduced:
```
NVIM v0.9.1
Build type: Release
LuaJIT 2.1.1692616192

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

### Additional context

To fix the repo after that, I had to do a git hard reset (soft or mixed resets didn't work).

git errors looked like that:
```
> git commit
error: invalid object 100644 11544a188ec3192eb3aab635520020d5f175bd0b for '.github/workflows/ci.yml'
fatal: unable to read 11544a188ec3192eb3aab635520020d5f175bd0b
> git status
fatal: unable to read 11544a188ec3192eb3aab635520020d5f175bd0b
```
(named file is from [tree-sitter-gitcommit](https://github.com/gbprod/tree-sitter-gitcommit))
(for experiment, I created it in the root of my repo, after which it started complaining about the next file in tree-sitter-gitcommit).

The test git repo where I reproduced the bug: [test.zip](https://github.com/nvim-treesitter/nvim-treesitter/files/12547471/test.zip)

**Comment by staticssleever668** [comment]
Just had a very similar with while installing "nvim-treesitter-textobjects", but I think this time it was a problem with lazy.nvim.

I added "nvim-treesitter-textobjects" as a dependency to lazy.nvim, exited Neovim.
Then I did a "git commit -p" in my repo. (I don't remember exactly how it happened as it caught me off-guard, so I *think*) I interactively added files, the editor opened and lazy.nvim installed nvim-treesitter-textobjects, I successfully commited.
Modified a file with some nvim-treesitter-textobjects actions, did a "git commit -p" again and saw this:
```
fatal: unable to read bb6a1423efc0537578e723900dadb14b36e554c6
Cannot close git diff-index --cached --numstat --summary HEAD -- () at /usr/lib/git-core/git-add--interactive line 183.
fatal: interactive add failed
```
Curiously, all my repo files were staged as "deleted", while some of nvim-treesitter-textobjects are staged as added and most are unstaged at all, chaos.
Luckily, since my first commit was successful and I was modifying my file in another Neovim instance, I did a "git reset --hard" and pressed "u" in that Neovim instance and got file modifications back.

Anyway, whether this is an issue with nvim-treesitter or lazy.nvim, I'll to make a better minimal reproducing example.

**Comment by mathieulj** [comment]
I've been debugging this issue all morning as I've run into it on my own machine with a fresh neovim install.

**Pre-requisites AFAICT for my reproduction:**

- Neovim with treesitter configured (a clean checkout of https://github.com/dam9000/kickstart-modular.nvim) does the trick for me
- Not installed nor locally cached treesitter plugin you wish to install
- git repository you have uncommitted changes in
- `require("nvim-treesitter.install").prefer_git = true` in your config
- Git editor is configured for nvim
- Git worktrees are in use (may not be required)
- ??

I don't believe these versions matter but just in case:

```
~/.config/nvim via 🌙 
❯ git --version
git version 2.42.0

~/.config/nvim via 🌙 
❯ nvim --version
NVIM v0.10.0
Build type: Release
LuaJIT 2.1.1713773202
Run "nvim -V1 -v" for more info

~/.config/nvim via 🌙 
❯ uname -a
Darwin Mathieus-MacBook-Pro-2.local 23.5.0 Darwin Kernel Version 23.5.0: Wed May  1 20:12:58 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6000 arm64
```

**Reproduction**

In any git repo, make a change and attempt to commit it:

```sh
mkdir /tmp/repo
cd /tmp/repo
git init .
touch somefile.txt
git add somefile.txt
git commit -m init
git worktree add -b new_branch /tmp/new_branch
cd /tmp/new_branch
touch otherfile.txt
git add otherfile.txt
git commit
```

If the gitcommit treesitter plugin didn't try to install automatically, manually try installing another plugin.

**Root cause**

When spawning the editor to create the commit message, git sets some environment variables:

```
GIT_DIR=/private/tmp/repo/.git/worktrees/new_branch
GIT_AUTHOR_DATE=@1717863612 -0400
GIT_INDEX_FILE=/private/tmp/repo/.git/worktrees/new_branch/index
GIT_AUTHOR_NAME=Mathieu Letendre-Jauniaux
GIT_PREFIX=
GIT_AUTHOR_EMAIL=mathieu.letendre-jauniaux@<work>.com
GIT_EXEC_PATH=/nix/store/q7avibx4c00h7zl3jfh3r5mc4xs6wsnj-git-2.42.0/libexec/git-core
```

Neovim will inherit these variables by default as will any process we spawn.

I'll have a PR up shortly with a fix.

**Comment by clason** [comment]
Thank you, but master is essentially frozen; we are completely rewriting the plugin on the `main` branch. Please test that first.

**Comment by mathieulj** [comment]
> And why are you setting prefer_git in the first place? That is just a bandaid for Windows where curl and tar may not be available.

I was not aware the feature should be avoided. To be fully honest, I had just cleared out my config and was starting from a template that had it configured. The docs here don't mention it being a bandaid, work in progress nor a windows only feature so I didn't see a reason to change it.

My work git monorepo was corrupted twice by this bug before I started digging into how/why it could be happening. If the feature is broken to the point where it will corrupt the repository the user is working on, should we not disable it, fix it, or at very least put up warning banners in the docs?
